### PR TITLE
GAWB-757 GET /entities now returns entity type counts and all the attribute names used by that type

### PR DESCRIPTION
**Pull request by helgridly**
GAWB-757.

Response:

```
{
    "sample" : {
        "count" : 5,
        "attributeNames" : ["entityName", "entityType", "tumortype"]
    },
    "pair" : {
        "count" : 2,
        "attributeNames" : ["entityName", "entityType", "case", "control"]
    },
    ...
}
```
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it **(apply requires_doge label)**
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala [hunk 1]
```diff
@@ -111,7 +113,7 @@ trait EntityComponent {
 
     }
 
-    def entityAttributes(entityId: Long) = for {
+    def entityAttributes(entityId: Rep[Long]): EntityAttributeQuery = for {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:116]: I needed to do this so I could call it from my for-comp. This function was unused until then anyway, and slick provides implicits for `Foo` to `Rep[Foo]` in most cases anyway.

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala [hunk 2]
```diff
@@ -81,6 +81,58 @@ class EntityComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers {
 
   }
 
+  it should "list all entity types with their attribute names" in withDefaultTestDatabase {
+
+    withWorkspaceContext(testData.workspace) { context =>
+
+      val desiredTypesAndAttrNames = Map(
+        "Sample" -> Seq("type", "whatsit", "thingies", "quot", "somefoo", "tumortype", "confused", "cycle"),
+        //"Aliquot" -> Seq(), NOTE: this is commented out because the db query doesn't return types that have no attributes.
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:90]: Is this okay? I think it's okay.

**Comment by MattPutnam** [comment]
👍 Exactly what I need.  Looks good, I'll defer to Doug for any Slick comments.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala [hunk 3]
```diff
@@ -293,6 +295,32 @@ trait EntityComponent {
...
 
+    def getAttrNamesAndEntityTypes(workspaceContext: SlickWorkspaceContext): ReadAction[Map[String, Seq[String]]] = {
+      val typesAndAttrNames = for {
+        entityRec <- filter(_.workspaceId === workspaceContext.workspaceId)
+        attrib <- entityAttributes(entityRec.id)
+      } yield {
+        (entityRec.entityType, attrib.name)
+      }
+
+      typesAndAttrNames.distinctOn(_._2).result map { result =>
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala]: doge says: did you mean: distinct?
yes, i did

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala (hunk 2 above)
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:90]: will this mean that Aliquot won't show up anywhere if there are entities but the entities have no attributes? that does not seem right.

Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:90]: In `getAttrNamesAndEntityTypes`, yes. But in the `getEntityTypeMetadata` call the keys get merged, so it'll be returned as an empty `Seq()` (as in the test for that function).

**Comment by helgridly** [comment]
I did the `distinctOn` -> `distinct` fix and added a test to catch it.

This is now waiting on a :+1: from @dvoet, specifically his opinion on whether [this](https://github.com/broadinstitute/rawls/pull/452/files#r65953534) needs fixing. I can directionally-join the entityType/attribute, I suppose.

**Comment by dvoet** [comment]
👍

**Pull request merged**